File: sdks/python-mcp/python/solvapay_mcp/oauth/auth_bridge.py
from __future__ import annotations

from collections.abc import Mapping

from solvapay_mcp.oauth.bearer import (
    extract_bearer_token,
    verify_bearer,
)


def build_auth_info_from_bearer(
    authorization: str | None,
    *,
    expected_issuer: str,
    expected_audience: str,
    now_unix_secs: int,
    jwks_json: object | None = None,
    hs256_secret: str | None = None,
    client_id: str | None = None,
    default_scopes: list[str] | None = None,
    include_payload: bool = False,
    claim_priority: list[str] | None = None,
) -> dict[str, object] | None:
    token = extract_bearer_token(authorization)
    if not token:
        return None
    result = verify_bearer(
        token,
        expected_issuer=expected_issuer,
        expected_audience=expected_audience,
        now_unix_secs=now_unix_secs,
        jwks_json=jwks_json,
        hs256_secret=hs256_secret,
        claim_priority=claim_priority,
    )
    if result.get("kind") != "ok":
        return None
    payload_raw = result.get("claims")
    payload: Mapping[str, object] = payload_raw if isinstance(payload_raw, dict) else {}
    customer_ref = result.get("customerRef")
    extra: dict[str, object] = {}
    if isinstance(customer_ref, str):
        extra["customer_ref"] = customer_ref
    resource = _resource(payload)
    if resource is not None:
        extra["resource"] = resource
    if include_payload:
        extra["payload"] = dict(payload)
    auth: dict[str, object] = {
        "token": token,
        "clientId": _client_id(payload, client_id),
        "scopes": _scopes(payload, default_scopes or []),
        "extra": extra,
    }
    expires_at = payload.get("exp")
    if isinstance(expires_at, int):
        auth["expiresAt"] = expires_at
    return auth


def _client_id(payload: Mapping[str, object], explicit: str | None) -> str:
    if explicit:
        return explicit
    for key in ("client_id", "azp"):
        value = payload.get(key)
        if isinstance(value, str) and value:
            return value
    return "solvapay-mcp-client"


def _resource(payload: Mapping[str, object]) -> str | None:
    for key in ("resource", "aud"):
        value = payload.get(key)
        if isinstance(value, str) and value:
            return value
    return None


def _scopes(payload: Mapping[str, object], default_scopes: list[str]) -> list[str]:
    scp = payload.get("scp")
    if isinstance(scp, list):
        return [item for item in scp if isinstance(item, str)]
    scope = payload.get("scope")
    if isinstance(scope, str) and scope.strip():
        return [part for part in scope.split() if part]
    return list(default_scopes)

File: sdks/typescript/mcp-core/src/auth-bridge.ts
/**
 * Build an MCP `authInfo` envelope from a verified `Authorization: Bearer`
 * JWT. Populates `authInfo.extra.customer_ref` so downstream
 * `getCustomerRef` extractors can read the caller identity.
 */

import { extractBearerToken, verifyBearer, type McpVerifyBearerOptions } from './bearer'
import type { McpToolExtra } from './types'

type JwtPayload = Record<string, unknown>

export type McpAuthInfoExtras = {
  clientId?: string
  defaultScopes?: string[]
  includePayload?: boolean
}

export interface BuildAuthInfoFromBearerOptions extends McpVerifyBearerOptions, McpAuthInfoExtras {}

function getClientId(payload: JwtPayload, explicitClientId?: string): string {
  if (explicitClientId) return explicitClientId

  const payloadClientId =
    (typeof payload.client_id === 'string' && payload.client_id) ||
    (typeof payload.azp === 'string' && payload.azp) ||
    null

  return payloadClientId || 'solvapay-mcp-client'
}

function getResource(payload: JwtPayload): string | undefined {
  if (typeof payload.resource === 'string' && payload.resource) return payload.resource
  if (typeof payload.aud === 'string' && payload.aud) return payload.aud
  return undefined
}

function getScopes(payload: JwtPayload, defaultScopes: string[]): string[] {
  if (Array.isArray(payload.scp)) {
    return payload.scp.filter(scope => typeof scope === 'string') as string[]
  }

  if (typeof payload.scope === 'string' && payload.scope.trim()) {
    return payload.scope
      .split(/\s+/)
      .map(scope => scope.trim())
      .filter(Boolean)
  }

  return defaultScopes
}

function getExpiresAt(payload: JwtPayload): number | undefined {
  return typeof payload.exp === 'number' ? payload.exp : undefined
}

export function buildAuthInfoFromBearer(
  authorization: string | null | undefined,
  options: BuildAuthInfoFromBearerOptions,
): McpToolExtra['authInfo'] | null {
  const token = extractBearerToken(authorization)
  if (!token) return null

  const verified = verifyBearer(token, options)
  if (verified.kind !== 'ok') {
    return null
  }
  const payload = verified.claims
  const clientId = getClientId(payload, options.clientId)
  const scopes = getScopes(payload, options.defaultScopes || [])
  const expiresAt = getExpiresAt(payload)
  const resource = getResource(payload)

  return {
    token,
    clientId,
    scopes,
    expiresAt,
    extra: {
      customer_ref: verified.customerRef,
      ...(resource ? { resource } : {}),
      ...(options.includePayload ? { payload } : {}),
    },
  }
}
